crowbar_framework/lib/crowbar/upgrade_status.rb

```diff
@@ -75,13 +75,19 @@ def current_step_state
       progress[:steps][current_step] || {}
     end
 
-    def start_step
+    # 'step' is name of the step user wants to start.
+    # If `out_of_order` is true, the step could be executed out of order
+    def start_step(step_name, out_of_order = false)
       ::Crowbar::Lock::LocalBlocking.with_lock(shared: false, logger: @logger, path: lock_path) do
-        if running?
+        if running? step_name
           @logger.warn("The step has already been started")
           return false
         end
-        progress[:steps][current_step][:status] = :running
+        if step_name != current_step && !out_of_order
```

Comment:
https://twitter.com/codinghorror/status/506010907021828096

In this case, "naming things" is the problem - [as I mentioned on 13th October](https://github.com/crowbar/crowbar-core/pull/692#issuecomment-253575454), this code will be pretty confusing for any newbies to read.  It is checking that the step the caller wants to start is the same as `current_step`, and then it is starting `current_step`.  So `current_step` doesn't mean "current step" at all!

It could be renamed to something like `next_or_current_step`. But even that's not quite right, because then the library is attempting to predict the future which is impossible because it has no idea what the next step will be: the user might want to proceed or repeat some steps.

A more correct name would be `current_or_next_pending_step`, but the length of that is a bad smell which highlights that the variable is trying to do two things at once:

1. track what is the next *pending* step (not necessarily the next one the user will take)
2. track what step is currently running, if any

Luckily we can fix this quite easily by splitting `current_step` into two:

1. `next_pending_step`
2. `current_step` - the current step, **or `nil` if there is no step running!!**

I am not suggesting to make those changes in this PR though - we can do a separate PR for that.